File: 134.java
import java.util.Arrays;

class Solution134 {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int sum = 0, ans = 0;
        // int[] arr = new int[cost.length];
        for (int i = 0; i < cost.length; i++) {
            arr[i] = gas[i] - cost[i];
            if (sum + gas[i] - cost[i] > gas[i] - cost[i]) {
                sum = sum + gas[i] - cost[i];
            } else
                sum = gas[i] - cost[i];
                ans = i;
        }
        
        // System.out.println(Arrays.toString(arr));
        return sum>=0?ans:-1;
    }

    public static void main(String[] args) {
        Solution134 s = new Solution134();
        int[] gas = { 1,2, 3, 4,5 }, cost = { 3, 4, 5,1,2 };
        System.out.println(s.canCompleteCircuit(gas, cost));
    }
}

File: 134.cpp
#include <vector>
#include <limits.h>
#include <iostream>
using namespace std;
class Solution
{
public:
    static int canCompleteCircuit(vector<int> &gas, vector<int> &cost)
    {
        int sum = INT_MAX;
        int temp = 0;
        int index = -1;
        for (size_t i = 0; i < gas.size()-1; i++)
        {
            temp += gas[i] - cost[i+1];
            if (temp < sum)     
            {
                sum = temp;
                index = i+1;
            }
            temp += gas[i];
        }
        temp = gas[index];
        int i = index+1;
        int n = 0;
        while (n++ != gas.size())
        {
            temp += gas[i] - cost[i];
            i=(i+1)%gas.size();
            if (temp < 0)
            {
                return -1;
            }
        }

        return index;
        }
};
int main(){
    vector<int> gas = {1, 2, 3, 4, 5};
    vector<int> cost = {3, 4, 5, 1, 2};
    cout << Solution::canCompleteCircuit(gas, cost)<<'\n';

}